Test case failures in Python 3.9

Following failures are noted in Python 3.9

```
PYTHONWARNINGS=always pytest                                      
/root/py39-venv/lib/python3.9/site-packages/_pytest/compat.py:340: PytestDeprecationWarning: The TerminalReporter.writer attribute is deprecated, use TerminalReporter._tw instead at your own risk.
See https://docs.pytest.org/en/stable/deprecations.html#terminalreporter-writer for more information.
  return getattr(object, name, default)
========================================================================= test session starts ==========================================================================
platform linux -- Python 3.9.0b5, pytest-6.0.1, py-1.9.0, pluggy-0.12.0
rootdir: /root/checked_repos/typish
plugins: hypothesis-4.43.1, cov-2.8.1
collected 120 items                                                                                                                                                    

tests/test_generic_collection_type.py ...                                                                                                                        [  2%]
tests/test_meta.py .                                                                                                                                             [  3%]
tests/classes/test_cls_dict.py ......                                                                                                                            [  8%]
tests/classes/test_cls_function.py ............                                                                                                                  [ 18%]
tests/classes/test_literal_meta.py ..                                                                                                                            [ 20%]
tests/classes/test_something.py .......                                                                                                                          [ 25%]
tests/classes/test_subscribtable_type.py ...                                                                                                                     [ 28%]
tests/decorators/test_hintable.py ...........................                                                                                                    [ 50%]
tests/functions/test_common_ancestor.py ......                                                                                                                   [ 55%]
tests/functions/test_get_args_and_return_type.py ....                                                                                                            [ 59%]
tests/functions/test_get_mro.py ...                                                                                                                              [ 61%]
tests/functions/test_get_type.py ..............                                                                                                                  [ 73%]
tests/functions/test_get_type_hints_of_callable.py .....                                                                                                         [ 77%]
tests/functions/test_instance_of.py ............F.                                                                                                               [ 89%]
tests/functions/test_is_type_annotation.py F                                                                                                                     [ 90%]
tests/functions/test_origin_and_alias.py ..                                                                                                                      [ 91%]
tests/functions/test_subclass_of.py ......F...                                                                                                                   [100%]

=============================================================================== FAILURES ===============================================================================
________________________________________________________________ TestInstanceOf.test_instance_of_union _________________________________________________________________

self = <tests.functions.test_instance_of.TestInstanceOf testMethod=test_instance_of_union>

/root/py39-venv/lib/python3.9/site-packages/pygments/util.py:27: DeprecationWarning: Flags not at the start of the expression '<(.+?)(\\s.*?)?>.*?</' (truncated)
  tag_re = re.compile(r'<(.+?)(\s.*?)?>.*?</.+?>(?uism)')
    def test_instance_of_union(self):
        self.assertTrue(instance_of(F(), Union[int, A]))
        self.assertTrue(instance_of(F(), Union[A, int]))
        self.assertTrue(instance_of(F(), Union[A, str]))
>       self.assertTrue(instance_of(F(), Optional[str]))
E       AssertionError: False is not true

tests/functions/test_instance_of.py:35: AssertionError
_____________________________________________________________ TestIsTypeAnnotation.test_is_type_annotation _____________________________________________________________

self = <tests.functions.test_is_type_annotation.TestIsTypeAnnotation testMethod=test_is_type_annotation>

    def test_is_type_annotation(self):
        # builtins and typing.
        self.assertTrue(is_type_annotation(int))
>       self.assertTrue(is_type_annotation(List))
E       AssertionError: False is not true

tests/functions/test_is_type_annotation.py:21: AssertionError
________________________________________________________________ TestSubclassOf.test_subclass_of_union _________________________________________________________________

self = <tests.functions.test_subclass_of.TestSubclassOf testMethod=test_subclass_of_union>

    def test_subclass_of_union(self):
        self.assertTrue(subclass_of(F, Union[int, str]))
        self.assertTrue(subclass_of(F, Union[A, int]))
        self.assertTrue(subclass_of(F, Union[A, B]))
        self.assertTrue(not subclass_of(int, Union[A, B]))
>       self.assertTrue(subclass_of(F, Optional[A]))
E       AssertionError: False is not true

tests/functions/test_subclass_of.py:70: AssertionError
======================================================================= short test summary info ========================================================================
FAILED tests/functions/test_instance_of.py::TestInstanceOf::test_instance_of_union - AssertionError: False is not true
FAILED tests/functions/test_is_type_annotation.py::TestIsTypeAnnotation::test_is_type_annotation - AssertionError: False is not true
FAILED tests/functions/test_subclass_of.py::TestSubclassOf::test_subclass_of_union - AssertionError: False is not true
==================================================================== 3 failed, 117 passed in 1.13s =====================================================================
```

--- a/.github/workflows/pythonapp.yml
+++ b/.github/workflows/pythonapp.yml
@@ -7,7 +7,7 @@ jobs:
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
-        python-version: [ '3.5', '3.6', '3.7', '3.8' ]
+        python-version: [ '3.5', '3.6', '3.7', '3.8', '3.9' ]
         os: [ ubuntu-latest, macOS-latest, windows-latest ]
     name: Python ${{ matrix.python-version }} on ${{ matrix.os }}
     steps:

--- a/.github/workflows/pythonapp.yml
+++ b/.github/workflows/pythonapp.yml
@@ -7,7 +7,7 @@ jobs:
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
-        python-version: [ '3.5', '3.6', '3.7', '3.8', '3.9' ]
+        python-version: [ '3.5', '3.6', '3.7', '3.8' ]
         os: [ ubuntu-latest, macOS-latest, windows-latest ]
     name: Python ${{ matrix.python-version }} on ${{ matrix.os }}
     steps:

--- a/setup.py
+++ b/setup.py
@@ -15,7 +15,7 @@ requirements = []
 
 test_requirements = [
     'numpy',
-    'nptyping>=1.1.0',
+    'nptyping>=1.3.0',
     'pycodestyle',
     'pylint',
     'mypy',

--- a/setup.py
+++ b/setup.py
@@ -56,5 +56,6 @@ setup(
         'Programming Language :: Python :: 3.6',
         'Programming Language :: Python :: 3.7',
         'Programming Language :: Python :: 3.8',
+        'Programming Language :: Python :: 3.9',
     ]
 )

--- a/tests/classes/test_literal_meta.py
+++ b/tests/classes/test_literal_meta.py
@@ -15,3 +15,11 @@ class TestLiteralMeta(TestCase):
 
     def test_str(self):
         self.assertEqual('Literal[42]', str(LiteralAlias[42]))
+
+    def test_multiple_args(self):
+        self.assertTrue(isinstance(1, LiteralAlias[1, 2]))
+        self.assertTrue(isinstance(2, LiteralAlias[1, 2]))
+        self.assertTrue(isinstance(1, LiteralAlias[(1, 2)]))
+        self.assertTrue(isinstance(2, LiteralAlias[(1, 2)]))
+        self.assertTrue(isinstance(1, LiteralAlias[((1, 2),)]))
+        self.assertTrue(isinstance(2, LiteralAlias[((1, 2),)]))

--- a/tests/classes/test_something.py
+++ b/tests/classes/test_something.py
@@ -84,6 +84,9 @@ class TestSomething(TestCase):
         self.assertEqual("typish.Something['x': int]", str(Something['x': int]))
         self.assertEqual("typish.Something['a': int, 'b': typing.Callable[[int, int], str]]",
                          repr(Inyerface))
+        self.assertEqual("typish.Something['x': ...]", str(Something['x': ...]))
+        self.assertEqual("typish.Something['x': C1]", str(Something['x': C1]))
+        self.assertEqual("typish.Something['x': b]", str(Something['x': C5.b]))
 
     def test_isinstance_generic_collection(self):
         isinstance(List[int], TypingType)

--- a/tests/functions/test_instance_of.py
+++ b/tests/functions/test_instance_of.py
@@ -99,6 +99,7 @@ class TestInstanceOf(TestCase):
         self.assertTrue(instance_of(Any, Literal[Any]))
         self.assertTrue(not instance_of(42, Literal[Any]))
         self.assertTrue(not instance_of(42, Literal))
+        self.assertTrue(instance_of(2, Literal[1, 2]))
 
     def test_instance_of_typing_literal(self):
         # This is to mock Python 3.8 Literal.

--- a/tests/functions/test_instance_of.py
+++ b/tests/functions/test_instance_of.py
@@ -1,5 +1,6 @@
+from sys import version_info
 from typing import List, Dict, Union, Optional, Callable, Any, Tuple, Type, Iterable
-from unittest import TestCase
+from unittest import TestCase, skipUnless
 
 import nptyping as nptyping
 import numpy
@@ -8,10 +9,20 @@ from typish import Literal, instance_of, State, register_get_type
 
 
 class A: pass
+
+
 class B(A): pass
+
+
 class C(B): pass
+
+
 class D(C): pass
+
+
 class E(D): pass
+
+
 class F(E, str): pass
 
 
@@ -128,3 +139,11 @@ class TestInstanceOf(TestCase):
         self.assertTrue(instance_of([arr], List[arr_type], state=local_state))
         self.assertTrue(instance_of([arr], List[nptyping.NDArray], state=local_state))
         self.assertTrue(not instance_of([arr], List[nptyping.NDArray[(4,), float]], state=local_state))
+
+    @skipUnless(10 * version_info.major + version_info.minor >= 39, 'PEP-585')
+    def test_instance_of_py39_types(self):
+        self.assertTrue(instance_of({'42': 42}, Dict[str, int]))
+        self.assertTrue(not instance_of({'42': 42}, Dict[str, str]))
+        self.assertTrue(not instance_of({42: 42}, Dict[str, int]))
+        self.assertTrue(instance_of([1, 2, 3], list[int]))
+        self.assertTrue(not instance_of([1, 2, 3], list[str]))

--- a/tests/functions/test_origin_and_alias.py
+++ b/tests/functions/test_origin_and_alias.py
@@ -16,6 +16,10 @@ from unittest import TestCase
 from typish import get_origin, get_alias
 
 
+class Union:
+    """To shadow typing.Union."""
+
+
 class TestOriginAndAlias(TestCase):
     def test_get_origin(self):
         self.assertEqual(list, get_origin(List[int]))
@@ -26,6 +30,7 @@ class TestOriginAndAlias(TestCase):
         self.assertEqual(defaultdict, get_origin(DefaultDict))
         self.assertEqual(type, get_origin(Type[int]))
         self.assertEqual(Set, get_origin(AbstractSet))
+        self.assertIn('test_origin_and_alias', str(get_origin(Union)))
 
     def test_get_alias(self):
         self.assertEqual(List, get_alias(list))

--- a/tests/functions/test_origin_and_alias.py
+++ b/tests/functions/test_origin_and_alias.py
@@ -20,6 +20,15 @@ class Union:
     """To shadow typing.Union."""
 
 
+class MetaMock(type):
+    __name__ = 'list'
+    __args__ = (str,)
+
+
+class ListMock(metaclass=MetaMock):
+    ...
+
+
 class TestOriginAndAlias(TestCase):
     def test_get_origin(self):
         self.assertEqual(list, get_origin(List[int]))
@@ -42,3 +51,8 @@ class TestOriginAndAlias(TestCase):
         self.assertEqual(DefaultDict, get_alias(defaultdict))
         self.assertEqual(Type, get_alias(type))
         self.assertEqual(AbstractSet, get_alias(Set))
+        self.assertEqual(List, get_alias(List))
+        self.assertEqual(Dict, get_alias(Dict))
+
+    def test_get_alias_from_parameterized_standard_list(self):
+        self.assertEqual(List[str], get_alias(ListMock))

--- a/typish/_meta.py
+++ b/typish/_meta.py
@@ -1,5 +1,5 @@
 __title__ = 'typish'
-__version__ = '1.7.0'
+__version__ = '1.8.0'
 __author__ = 'Ramon Hagenaars'
 __author_email__ = 'ramon.hagenaars@gmail.com'
 __description__ = 'Functionality for types'

--- a/typish/classes/_literal.py
+++ b/typish/classes/_literal.py
@@ -33,7 +33,7 @@ class _LiteralMeta(SubscriptableType):
                     result = result[0]  # result was a tuple in a tuple.
                 if result and not isinstance(result, tuple):
                     result = (result,)
-            except AttributeError:
+            except AttributeError:  # pragma: no cover
                 # In case of Python 3.5
                 result = tuple()
         elif item in ('__origin__', '__name__', '_name'):

--- a/typish/classes/_literal.py
+++ b/typish/classes/_literal.py
@@ -27,7 +27,12 @@ class _LiteralMeta(SubscriptableType):
         """
         if item == '__args__':
             try:
-                result = SubscriptableType.__getattribute__(cls, item),
+                result = SubscriptableType.__getattribute__(cls, item)
+                if (result and isinstance(result, tuple)
+                        and isinstance(result[0], tuple)):
+                    result = result[0]  # result was a tuple in a tuple.
+                if result and not isinstance(result, tuple):
+                    result = (result,)
             except AttributeError:
                 # In case of Python 3.5
                 result = tuple()
@@ -38,10 +43,11 @@ class _LiteralMeta(SubscriptableType):
         return result
 
     def __instancecheck__(self, instance):
-        return self.__args__ and self.__args__[0] == instance
+        return self.__args__ and instance in self.__args__
 
     def __str__(self):
-        return '{}[{}]'.format(self.__name__, self.__args__[0])
+        args = ', '.join(str(arg) for arg in self.__args__)
+        return '{}[{}]'.format(self.__name__, args)
 
     def __subclasscheck__(self, subclass: typing.Any) -> bool:
         return is_literal_type(subclass)
@@ -61,7 +67,7 @@ class LiteralAlias(type, metaclass=_LiteralMeta):
         from typish.functions._get_args import get_args
 
         args = get_args(literal)
-        return LiteralAlias[args[0]] if args else LiteralAlias
+        return LiteralAlias[args] if args else LiteralAlias
 
 
 # If Literal is available (Python 3.8+), then return that type instead.

--- a/typish/classes/_something.py
+++ b/typish/classes/_something.py
@@ -7,7 +7,7 @@ from typish.classes._subscriptable_type import SubscriptableType
 
 class _SomethingMeta(SubscriptableType):
     """
-    This metaclass is coupled to ``Interface``.
+    This metaclass is coupled to ``Something``.
     """
     def __instancecheck__(self, instance: object) -> bool:
         # Check if all attributes from self.signature are also present in
@@ -66,9 +66,7 @@ class _SomethingMeta(SubscriptableType):
         else, we fall back on repr(obj).
         """
         if isinstance(obj, type) and not issubclass(obj, Callable):
-            if obj.__module__ == 'builtins':
-                return obj.__qualname__
-            return '{}.{}'.format(obj.__module__, obj.__qualname__)
+            return obj.__qualname__
         if obj is ...:
             return '...'
         if isinstance(obj, types.FunctionType):
@@ -104,7 +102,7 @@ class Something(type, metaclass=_SomethingMeta):
     def __getattr__(cls, item):
         # This method exists solely to fool the IDE into believing that
         # Something can have any attribute.
-        return type.__getattr__(cls, item)
+        return type.__getattr__(cls, item)  # pragma: no cover
 
     @staticmethod
     def like(obj: Any, exclude_privates: bool = True) -> 'Something':

--- a/typish/functions/_get_alias.py
+++ b/typish/functions/_get_alias.py
@@ -1,6 +1,8 @@
 import typing
 from functools import lru_cache
 
+from typish.functions._is_from_typing import is_from_typing
+
 from typish._types import T
 
 
@@ -8,12 +10,21 @@ from typish._types import T
 def get_alias(cls: T) -> typing.Optional[T]:
     """
     Return the alias from the ``typing`` module for ``cls``. For example, for
-    ``cls=list``, the result would be ``typing.List``. If no alias exists for
-    ``cls``, then ``None`` is returned.
+    ``cls=list``, the result would be ``typing.List``. If ``cls`` is
+    parameterized (>=3.9), then a parameterized ``typing`` equivalent is
+    returned. If no alias exists for ``cls``, then ``None`` is returned.
+    If ``cls`` already is from ``typing`` it is returned as is.
     :param cls: the type for which the ``typing`` equivalent is to be found.
     :return: the alias from ``typing``.
     """
-    return _alias_per_type.get(cls.__name__, None)
+    if is_from_typing(cls):
+        return cls
+    alias = _alias_per_type.get(cls.__name__, None)
+    if alias:
+        args = getattr(cls, '__args__', tuple())
+        if args:
+            alias = alias[args]
+    return alias
 
 
 _alias_per_type = {

--- a/typish/functions/_get_origin.py
+++ b/typish/functions/_get_origin.py
@@ -1,6 +1,9 @@
 import typing
 from collections import deque, defaultdict
 from collections.abc import Set
+from inspect import isclass
+
+from typish.functions._is_from_typing import is_from_typing
 
 
 def get_origin(t: type) -> type:
@@ -14,7 +17,9 @@ def get_origin(t: type) -> type:
 
     simple_name = get_simple_name(t)
     result = _type_per_alias.get(simple_name, None)
-    if not result:
+    if isclass(t) and not is_from_typing(t):
+        result = t
+    elif not result:
         result = getattr(typing, simple_name, t)
     return result
 

--- a/typish/functions/_get_origin.py
+++ b/typish/functions/_get_origin.py
@@ -29,4 +29,5 @@ _type_per_alias = {
     'DefaultDict': defaultdict,
     'Type': type,
     'AbstractSet': Set,
+    'Optional': typing.Union,
 }

--- a/typish/functions/_instance_of.py
+++ b/typish/functions/_instance_of.py
@@ -1,7 +1,7 @@
 from typish._state import DEFAULT, State
 
 
-def instance_of(obj: object, *args: type, state: State = DEFAULT) -> bool:
+def instance_of(obj: object, *args: object, state: State = DEFAULT) -> bool:
     """
     Check whether ``obj`` is an instance of all types in ``args``, while also
     considering generics.
@@ -18,7 +18,7 @@ def instance_of(obj: object, *args: type, state: State = DEFAULT) -> bool:
     return all(_instance_of(obj, clsinfo, state) for clsinfo in args)
 
 
-def _instance_of(obj: object, clsinfo: type, state: State = DEFAULT) -> bool:
+def _instance_of(obj: object, clsinfo: object, state: State = DEFAULT) -> bool:
     from typish.classes._literal import LiteralAlias, is_literal_type
     from typish.functions._subclass_of import subclass_of
     from typish.functions._get_type import get_type

--- a/typish/functions/_is_type_annotation.py
+++ b/typish/functions/_is_type_annotation.py
@@ -18,4 +18,5 @@ def is_type_annotation(item: typing.Any) -> bool:
     return (item is typing.Any
             or instance_of(item, type)
             or instance_of(item, super_cls)
+            or getattr(item, '__module__', None) == 'typing'
             or isinstance(item, UnionType))

--- a/typish/functions/_is_type_annotation.py
+++ b/typish/functions/_is_type_annotation.py
@@ -15,7 +15,9 @@ def is_type_annotation(item: typing.Any) -> bool:
     # Use _GenericAlias for Python 3.7+ and use GenericMeta for the rest.
     super_cls = getattr(typing, '_GenericAlias',
                         getattr(typing, 'GenericMeta', None))
-    return (item is typing.Any
+
+    return not isinstance(item, typing.TypeVar) and (
+            item is typing.Any
             or instance_of(item, type)
             or instance_of(item, super_cls)
             or getattr(item, '__module__', None) == 'typing'

--- a/typish/functions/_subclass_of.py
+++ b/typish/functions/_subclass_of.py
@@ -1,6 +1,7 @@
 import typing
 
 from typish._types import Unknown
+from typish.functions._get_alias import get_alias
 
 
 def subclass_of(cls: object, *args: object) -> bool:
@@ -21,6 +22,9 @@ def subclass_of(cls: object, *args: object) -> bool:
 def _subclass_of(cls: type, clsinfo: object) -> bool:
     # Check whether cls is a subtype of clsinfo.
     from typish.classes._literal import LiteralAlias
+
+    # Translate to typing type if possible.
+    clsinfo = get_alias(clsinfo) or clsinfo
 
     if _is_true_case(cls, clsinfo):
         result = True
@@ -87,7 +91,7 @@ def _subclass_of_generic(
         matched_info_args = info_args * len(args)
         result = _subclass_of_tuple(args, matched_info_args)
     elif (subclass_of(cls_origin, info_generic_type) and cls_args
-            and len(cls_args) == len(info_args)):
+          and len(cls_args) == len(info_args)):
         result = all(subclass_of(*tup) for tup in zip(cls_args, info_args))
     # Note that issubtype(list, List[...]) is always False.
     # Note that the number of arguments must be equal.

--- a/typish/functions/_subclass_of.py
+++ b/typish/functions/_subclass_of.py
@@ -36,7 +36,7 @@ def _subclass_of(cls: type, clsinfo: type) -> bool:
 
 def _forward_subclass_check(cls: type, clsinfo: type) -> bool:
     # Forward the subclass check for cls and clsinfo to delegates that know how
-    # to check that particulat cls/clsinfo type.
+    # to check that particular cls/clsinfo type.
 
     from typish.functions._get_origin import get_origin
     from typish.functions._get_args import get_args

--- a/typish/functions/_subclass_of.py
+++ b/typish/functions/_subclass_of.py
@@ -3,7 +3,7 @@ import typing
 from typish._types import Unknown
 
 
-def subclass_of(cls: type, *args: type) -> bool:
+def subclass_of(cls: object, *args: object) -> bool:
     """
     Return whether ``cls`` is a subclass of all types in ``args`` while also
     considering generics.
@@ -18,7 +18,7 @@ def subclass_of(cls: type, *args: type) -> bool:
     return all(_subclass_of(cls, clsinfo) for clsinfo in args)
 
 
-def _subclass_of(cls: type, clsinfo: type) -> bool:
+def _subclass_of(cls: type, clsinfo: object) -> bool:
     # Check whether cls is a subtype of clsinfo.
     from typish.classes._literal import LiteralAlias
 

